programs/cleos/do_http_post_libcurl.cpp

```diff
@@ -0,0 +1,142 @@
+      size_t       realsize = size * nmemb;
+      std::string& mem      = *(std::string*)userp;
+      mem.append((const char*)data, realsize);
+      return realsize;
+   }
+
+   // the following two functions are directly copy from the libcurl API overview from
+   // https://curl.se/libcurl/c/CURLOPT_DEBUGFUNCTION.html
+   void dump(const char* text, FILE* stream, unsigned char* ptr, size_t size) {
+      size_t       i;
+      size_t       c;
+      unsigned int width = 0x10;
+
+      fprintf(stream, "%s, %10.10ld bytes (0x%8.8lx)\n", text, (long)size, (long)size);
+
+      for (i = 0; i < size; i += width) {
+         fprintf(stream, "%4.4lx: ", (long)i);
+
+         /* show hex to the left */
+         for (c = 0; c < width; c++) {
+            if (i + c < size)
+               fprintf(stream, "%02x ", ptr[i + c]);
+            else
+               fputs("   ", stream);
+         }
+
+         /* show data on the right */
+         for (c = 0; (c < width) && (i + c < size); c++) {
+            char x = (ptr[i + c] >= 0x20 && ptr[i + c] < 0x80) ? ptr[i + c] : '.';
+            fputc(x, stream);
+         }
+
+         fputc('\n', stream); /* newline */
+      }
+   }
+
+   int my_trace(CURL* handle, curl_infotype type, char* data, size_t size, void* userp) {
+      const char* text;
+      (void)handle; /* prevent compiler warning */
+      (void)userp;
+
+      switch (type) {
+         case CURLINFO_TEXT: fprintf(stderr, "== Info: %s", data);
+         default: /* in case a new one is introduced to shock us */ return 0;
+
+         case CURLINFO_HEADER_OUT: text = "=> Send header"; break;
+         case CURLINFO_DATA_OUT: text = "=> Send data"; break;
+         case CURLINFO_SSL_DATA_OUT: text = "=> Send SSL data"; break;
+         case CURLINFO_HEADER_IN: text = "<= Recv header"; break;
+         case CURLINFO_DATA_IN: text = "<= Recv data"; break;
+         case CURLINFO_SSL_DATA_IN: text = "<= Recv SSL data"; break;
+      }
+
+      dump(text, stderr, (unsigned char*)data, size);
+      return 0;
+   }
+
+   struct curl_handle : std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> {
+      curl_handle(int) : std::unique_ptr<CURL, decltype(&curl_easy_cleanup)>( curl_easy_init(), &curl_easy_cleanup) {}
+   };
+
+
+   std::tuple<unsigned int, std::string> do_http_post(const std::string& base_uri, const std::string& path,
+                                                      const std::vector<std::string>& headers,
+                                                      const std::string& postjson, bool verify_cert, bool verbose,
+                                                      bool trace) {
+
+      static bool initialized = false;
+
+      if (!initialized) {
+         auto res = curl_global_init(CURL_GLOBAL_DEFAULT);
+         EOS_ASSERT(res == CURLE_OK, chain::http_exception, curl_easy_strerror(res));
+         initialized = true;
+      }
+
+      static std::map<std::string, curl_handle> handles;
```

Comment:
Is this map necessary? The documentation seems to suggest doing this is not required.

Reply:
According to https://everything.curl.dev/libcurl/connectionreuse:

> When you are using the easy API, or, more specifically, curl_easy_perform(), libcurl will keep the pool associated with the specific easy handle. Then reusing the SAME easy handle will ensure it can reuse its connection.

Reply:
> When libcurl is about to make a new connection for the purposes of doing a transfer, it will first check to see if there is an existing connection in the pool that it can reuse instead. The connection re-use check is done before any DNS or other name resolving mechanism is used, so it is purely host name based. If there is an existing live connection to the right host name, a lot of other properties (port number, protocol, etc) are also checked to see that it can be used.

Reply:
What's wrong to go with the public API contract I cited? Even if what you said is true that the cleanup is not necessary, I still prefer to have them. In this case, it becomes coding style preference because the performance is not an issue here. Similarly, when I am writing EOS contract code which links to simple allocator, I will still make sure all the memory allocated would be freed in my code even though I know perfectly that the free is actually no-op under the circumstance. It's a coding guideline I would always adhere to.

Reply:
There is nothing on that page that indicates it's the user's responsibility to strictly use the same hosts with a single handle, or that mixing hosts to a single handle would invalidate connection reuse. i.e. the documentation suggests it is sufficient to keep around a single curl handle and reuse it even when mixing different hosts (or protocols, etc).